node/node.go

```diff
@@ -0,0 +1,310 @@
+			errors.New("dev EUI must be 8 bytes"))
+	}
+	if conf.AppKey == "" {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("app key is required for OTAA join type"))
+	}
+	if len(conf.AppKey) != 32 {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("app key must be 16 bytes"))
+	}
+	return nil, nil
+}
+
+func (conf *Config) validateABPAttributes(path string) ([]string, error) {
+	if conf.AppSKey == "" {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("app session key is required for ABP join type"))
+	}
+	if len(conf.AppSKey) != 32 {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("app session key must be 16 bytes"))
+	}
+	if conf.NwkSKey == "" {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("network session key is required for ABP join type"))
+	}
+	if len(conf.NwkSKey) != 32 {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("network session key must be 16 bytes"))
+	}
+	if conf.DevAddr == "" {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("device address is required for ABP join type"))
+	}
+	if len(conf.DevAddr) != 8 {
+		return nil, resource.NewConfigValidationError(path,
+			errors.New("device address must be 4 bytes"))
+	}
+
+	return nil, nil
+
+}
+
+type Node struct {
+	resource.Named
+	logger logging.Logger
+
+	nwkSKey []byte
+	AppSKey []byte
+	AppKey  []byte
+
+	Addr   []byte
+	DevEui []byte
+
+	DecoderPath string
+	NodeName    string
+	gateway     sensor.Sensor
+}
+
+func newNode(
+	ctx context.Context,
+	deps resource.Dependencies,
+	conf resource.Config,
+	logger logging.Logger,
+) (sensor.Sensor, error) {
+	n := &Node{
+		Named:    conf.ResourceName().AsNamed(),
+		logger:   logger,
+		NodeName: conf.ResourceName().AsNamed().Name().Name,
+	}
+
+	cfg, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return nil, err
+	}
+
+	switch cfg.JoinType {
+	case "OTAA", "":
+		appKey, err := hex.DecodeString(cfg.AppKey)
+		if err != nil {
+			return nil, err
+		}
+		n.AppKey = appKey
+
+		devEui, err := hex.DecodeString(cfg.DevEUI)
+		if err != nil {
+			return nil, err
+		}
+		n.DevEui = devEui
+	case "ABP":
+		devAddr, err := hex.DecodeString(cfg.DevAddr)
+		if err != nil {
+			return nil, err
+		}
+
+		n.Addr = devAddr
+
+		appSKey, err := hex.DecodeString(cfg.AppSKey)
+		if err != nil {
+			return nil, err
+		}
+
+		n.AppSKey = appSKey
+	}
+
+	n.DecoderPath = cfg.DecoderPath
+
+	gateway, err := getGateway(ctx, deps)
+	if err != nil {
+		return nil, err
+	}
+
+	cmd := make(map[string]interface{})
+
+	// send the device to the gateway.
+	cmd["register_device"] = n
+
+	_, err = gateway.DoCommand(ctx, cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	n.gateway = gateway
+
+	return n, nil
+}
+
+func (n *Node) Reconfigure(ctx context.Context, deps resource.Dependencies, conf resource.Config) error {
+	cfg, err := resource.NativeConfig[*Config](conf)
```

Comment:
Can you just call the  reconfigure function in the constructor, or have this resource be always rebuild, they seem to be doing the same thing in terms of populating the gateway struct?

Reply:
oops yeah I meant to remove this, I want to just do always rebuild